Filtered_kernel/include/CGAL/Filtered_predicate.h

```diff
@@ -111,6 +113,78 @@ Filtered_predicate<EP,AP,C2E,C2A,Protection>::
     return ep(c2e(args)...);
 }
 
+template <class EP_RT, class EP_FT, class AP, class C2E_RT, class C2E_FT, class C2A, bool Protection = true>
+class Filtered_predicate_RT_FT
+{
+  C2E_RT c2e_rt;
+  C2E_FT c2e_ft;
+  C2A c2a;
+  EP_RT ep_rt;
+  EP_FT ep_ft;
+  AP ap;
+
+  using Ares = typename AP::result_type;
+
+public:
+  using result_type =  typename EP_FT::result_type;
+
+  template <typename... Args>
+  struct Call_operator_needs_FT
+  {
+    // This type traits class checks if the call operator can be called with
+    // `(const Args&..., FT_necessary())`.
+    using ArrayOfOne = char[1];
+    using ArrayOfTwo = char[2];
+
+    static ArrayOfOne& test(...);
+
+    template <typename... Args2>
+    static auto test(const Args2 &...args)
+        -> decltype(ap(c2a(args)..., FT_necessary()),
+                    std::declval<ArrayOfTwo &>());
+
+    enum { value = sizeof(test(std::declval<const Args&>()...)) == sizeof(ArrayOfTwo) };
+  };
+
+  // ## Important note
+  //
+  // If you want to remove of rename that member function template `needs_FT`,
+  // please also change the lines with
+  // `CGAL_GENERATE_MEMBER_DETECTOR(needs_FT);`
+  // or `has_needs_FT<typename R::Compare_distance_3>` in
+  // the file `Kernel_23/test/Kernel_23/include/CGAL/_test_new_3.h`.
+  template <typename... Args>
+  constexpr bool needs_FT(const Args&...) const {
+    return Call_operator_needs_FT<Args...>::value;
+  }
+
+  template <typename... Args>
+  result_type
+  operator()(const Args&... args) const
+  {
+    CGAL_BRANCH_PROFILER(std::string(" failures/calls to   : ") + std::string(CGAL_PRETTY_FUNCTION), tmp);
+    // Protection is outside the try block as VC8 has the CGAL_CFG_FPU_ROUNDING_MODE_UNWINDING_VC_BUG
+    {
+      Protect_FPU_rounding<Protection> p;
+      try
+        {
+          Ares res = ap(c2a(args)...);
+          if (is_certain(res))
+            return get_certain(res);
+        }
+      catch (Uncertain_conversion_exception&) {}
+    }
+    CGAL_BRANCH_PROFILER_BRANCH(tmp);
+    Protect_FPU_rounding<!Protection> p(CGAL_FE_TONEAREST);
+    CGAL_expensive_assertion(FPU_get_cw() == CGAL_FE_TONEAREST);
+    if constexpr (Call_operator_needs_FT<Args...>::value)
```

Comment:
`C:\CGAL_ROOT\CGAL-5.6-Ic-80\include\CGAL/Filtered_predicate.h(180): error C4984: 'if constexpr' is a C++17 language extension [C:\CGAL_ROOT\CGAL-5.6-Ic-80\cmake\platforms\MSVC2017-Debug-64bits\test\Generator\generic_random_test.vcxproj]`